text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+
+Between these two projects, the majority of the goals i have for this project
+have already been demonstrated to be feasible; the challenge here will be
+largely to figure out a way to combine them in a sensible and ergonomic way.
+
+Outside of Rust, I think it will also be important to spend some quality time
+with HarfBuzz, to make sure there are not requirements of that use-case that I
+am overlooking.
+
+
+# Schedule and time allocation
+
+The initial goal for this work is to determine whether it is feasible at all.
+With that in mind, I think we can break the project up into three phases, where
+we can abandon the work if any particular phase is a failure:
+
+
+1. macros, zero-copy types: demonstrate that we can use macros to generate code
+   for interpreting and accessing values from a small handful of tables,
+   including some tricky tables like GPOS/GSUB. 4-6 weeks.
+2. mutability: decide on a strategy for mutability, and implement it for that
+   same handful of tables. 6-10 weeks. It is possible that as we work on
+   mutability we will decide that we need to revisit how we structure our
+   code-gen code: for instance if we need to generate multiple types for each
+   table, a procedural macro may not be the best choice.
+3. Implement all core tables: ideally this is mostly just a matter of typing,
+   but it might still be quite a lot of typing. I’ll feel more comfortable
+   giving this a time estimate after I’ve finished the previous items, but if I
+   had to guess I would say ~12 weeks.
+
+
+# Questions
+
+There are a few remaining things I am currently unsure about:
+
+- what tables have the weirdest edge cases / pose the biggest potential
+  challenges, either for reading or writing? For instance serializing `loca` may
+  depends on contents of `glyf`?
+- how should the library handle errors during parsing? Is there any utility in
+  reporting the locations of errors? Do we assume good data? If we try to handle
+  errors ‘correctly’ then likely the entire API surface will need to return
+  Result types, and that might get exhausting.
+- is there a DRY method for generating methods that access various fields on
+  various tables? How much variation is there likely to be in terms of the API
+  requirements of different tables?
+- How extensively do we want to be using concrete types, in place of things like
+  type aliases/raw scalars? For instance a u16 can be used to represent a large
+  number of possible things in the spec, from a glyph identifier to an
+  F2DOT14 to an Offset16, an FWORD, etc etc. My inclination would be to have
+  newtypes for all distinct data types, but that might seem cumbersome to some
+  folks.
+- Certain collections are heterogeneous, such as the different subtable types in a
+  layout table. Can we generate nice code that represents these types using
+  enums where appropriate?
+
+
+
+[fonttools-rs]: https://github.com/simoncozens/fonttools-rs
```

Comment:
HarfBuzz and FontTools seem like they should be cited as well? - despite not being Rust they are the key things doing this work today.